core/drivers/imx_rngb.c

```diff
@@ -0,0 +1,218 @@
+#include <tee/tee_cryp_utl.h>
+#include <trace.h>
+#include <util.h>
+
+#define RNGB_COMMAND			0x0004
+#define RNGB_CONTROL			0x0008
+#define RNGB_STATUS			0x000C
+#define RNGB_ERROR			0x0010
+#define RNGB_FIFO			0x0014
+
+#define RNGB_CMD_CLR_ERR		0x00000020
+#define RNGB_CMD_CLR_INT		0x00000010
+#define RNGB_CMD_SEED			0x00000002
+#define RNGB_CMD_SELF_TEST		0x00000001
+
+#define RNGB_CTRL_MASK_ERROR		0x00000040
+#define RNGB_CTRL_MASK_DONE		0x00000020
+
+#define RNGB_STATUS_ERROR		0x00010000
+#define RNGB_STATUS_FIFO_LEVEL_MASK	0x00000f00
+#define RNGB_STATUS_FIFO_LEVEL_SHIFT	8
+#define RNGB_STATUS_SEED_DONE		0x00000020
+#define RNGB_STATUS_ST_DONE		0x00000010
+
+#define RNGB_ERROR_STATUS_STAT_ERR	0x00000008
+
+static struct imx_rng {
+	struct io_pa_va base;
+	uint32_t error;
+} rngb = {
+#if !defined(CFG_DT) || defined(CFG_EXTERNAL_DTB_OVERLAY)
+#if defined(RNGB_BASE)
+	.base.pa = RNGB_BASE,
+#else
+#error IMX_RNGB driver not supported on this platform
+#endif
+#endif
+};
+
+static void wait_for_irq(struct imx_rng *rng)
+{
+	uint64_t tref = timeout_init_us(1000000);
+	uint32_t status = 0;
+
+	while (!timeout_elapsed(tref)) {
+		rng->error = io_read32(rng->base.va + RNGB_ERROR);
+		status = io_read32(rng->base.va + RNGB_STATUS);
+		if (status & (RNGB_STATUS_SEED_DONE | RNGB_STATUS_ST_DONE))
+			return;
+	}
+	panic();
+}
+
+static void irq_clear(struct imx_rng *rng)
+{
+	uint32_t ctrl = 0;
+	uint32_t cmd = 0;
+
+	ctrl = io_read32(rng->base.va + RNGB_CONTROL);
+	ctrl |= RNGB_CTRL_MASK_DONE | RNGB_CTRL_MASK_ERROR;
+	io_write32(rng->base.va + RNGB_CONTROL, ctrl);
+
+	cmd = io_read32(rng->base.va + RNGB_COMMAND);
+	cmd |= RNGB_CMD_CLR_INT | RNGB_CMD_CLR_ERR;
+	io_write32(rng->base.va + RNGB_COMMAND, cmd);
+}
+
+static void irq_unmask(struct imx_rng *rng)
+{
+	uint32_t ctrl = 0;
+
+	ctrl = io_read32(rng->base.va + RNGB_CONTROL);
+	ctrl &= ~(RNGB_CTRL_MASK_DONE | RNGB_CTRL_MASK_ERROR);
+	io_write32(rng->base.va + RNGB_CONTROL, ctrl);
+}
+
+static void rng_seed(struct imx_rng *rng)
+{
+	uint64_t tref = timeout_init_us(1000000);
+	uint32_t cmd = 0;
+
+	cmd = io_read32(rng->base.va + RNGB_COMMAND);
+	io_write32(rng->base.va + RNGB_COMMAND, cmd | RNGB_CMD_CLR_ERR);
+	do {
+		irq_unmask(rng);
+		cmd = io_read32(rng->base.va + RNGB_COMMAND);
+		io_write32(rng->base.va + RNGB_COMMAND, cmd | RNGB_CMD_SEED);
+		wait_for_irq(rng);
+		irq_clear(rng);
+		if (!rng->error)
+			return;
+	} while (!timeout_elapsed(tref));
+	panic();
+}
+
+#if defined(CFG_DT) && !defined(CFG_EXTERNAL_DTB_OVERLAY)
+static const char *const rng_match = "fsl,imx25-rngb";
+static TEE_Result map_controller(void)
+{
+	void *fdt = get_dt();
+	size_t size = 0;
+	int off = 0;
+
+	if (!fdt)
+		return TEE_ERROR_NOT_SUPPORTED;
+
+	off = fdt_node_offset_by_compatible(fdt, off, rng_match);
+	if (off < 0)
+		return TEE_ERROR_NOT_SUPPORTED;
+
+	if (dt_map_dev(fdt, off, &rngb.base.va, &size) < 0)
+		return TEE_ERROR_NOT_SUPPORTED;
+
+	rngb.base.pa = virt_to_phys((void *)rngb.base.va);
+
```

Comment:
that is part of a more generic reservation commit that I want to rework and propose again
https://github.com/OP-TEE/optee_os/pull/4277

did you have a proposal?

Reply:
What we do is that we boot with SPL method so that we can dynamically adjust the system based in HWID information. Then load IMX header wrapped FIT image for u-boot that has op-tee and u-boot DTB as a payload. u-boot DTB is given to OP-TEE also. OP-TEE uses that and injects information. Then in u-boot we carry out op-tee's injected DT information to kernel DT.

When I last time looked the u-boot upstream it didn't have the SPL booting for imx6solox so we could try to push fixes for that (ROM problem workaround). imx6ull has been in backburner for a while but we also had SPL booting flow for that. I suppose some of the fixes we have we should be sending them to NXP as not everything is yet in upstream repos.

Reply:
yep, we do the same (also on imx8). But also adding overlays (not just the dtb)

Reply:
Yes, I would have the driver doing the reservation by modifying the device tree. Just like the DCP driver with ```dt_enable_secure_status()```. So yes, I'm totally fine with it.
With the default NXP bootflow, you've got U-boot loading the DTB in memory before booting OPTEE. From there, you can modify the DTB directly in memory.

NXP does not support the SPL booting for imx6/7 platforms.

If you want to go with SPL bootflow, you might want to use a DTB overlay where OPTEE creates a DTB overlay or appends an already existing one in memory. Then you would have the DTB and the overlay merged in the next boot stage ; typically U-boot.

Reply:
right, at Foundries, we do support SPL boot on imx6/7/8. 

Is there a reason why NXP doesnt?

Reply:
Just a comment why we are doing SPL way is as that longevity of our products is quite long -- thus we face component EOL's quite often as an example DDR memory might go EOL during device life time 2 or more times. When we do such updated hardware we increment HWID so that we can support multiple hardware revisions with same firmware.

We would be more than happy to share our stuff with NXP if that would help NXP to support also SPL method.

Reply:
@ldts I guess that NXP does not support it because it requires some extra re-work/maintenance.
Even if imx6/7 is still maintained, the focus has shifted to newer i.MX platforms.